### ISSUE TITLE
npm install error

### ISSUE THREAD

**Issue by SuperArchN** [nodejs/help#3608]
### Details

I have previously installed node.js on my mac book (2013) wit hno issues. But on my macbookpro (m1) I can't get npm install to work without a ton of errors. If anyone can give me any help i would really appreciate it.

### Node.js version

16.13.0

### Example code

arn@As-MacBook-Pro hashlips_art_engine-main % npm install
npm ERR! code 1
npm ERR! path /Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas
npm ERR! command failed
npm ERR! command sh -c node-pre-gyp install --fallback-to-build
npm ERR! Failed to execute '/usr/local/bin/node /usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --module=/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas/build/Release/canvas.node --module_name=canvas --module_path=/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas/build/Release --napi_version=8 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v93' (1)
npm ERR! node-pre-gyp info it worked if it ends with ok
npm ERR! node-pre-gyp info using node-pre-gyp@1.0.6
npm ERR! node-pre-gyp info using node@16.13.0 | darwin | arm64
npm ERR! node-pre-gyp info check checked for "/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas/build/Release/canvas.node" (not found)
npm ERR! node-pre-gyp http GET https://github.com/Automattic/node-canvas/releases/download/v2.8.0/canvas-v2.8.0-node-v93-darwin-unknown-arm64.tar.gz
npm ERR! node-pre-gyp ERR! install response status 404 Not Found on https://github.com/Automattic/node-canvas/releases/download/v2.8.0/canvas-v2.8.0-node-v93-darwin-unknown-arm64.tar.gz 
npm ERR! node-pre-gyp WARN Pre-built binaries not installable for canvas@2.8.0 and node@16.13.0 (node-v93 ABI, unknown) (falling back to source compile with node-gyp) 
npm ERR! node-pre-gyp WARN Hit error response status 404 Not Found on https://github.com/Automattic/node-canvas/releases/download/v2.8.0/canvas-v2.8.0-node-v93-darwin-unknown-arm64.tar.gz 
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@8.2.0
npm ERR! gyp info using node@16.13.0 | darwin | arm64
npm ERR! gyp info ok 
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@8.2.0
npm ERR! gyp info using node@16.13.0 | darwin | arm64
npm ERR! gyp info find Python using Python version 3.8.2 found at "/Library/Developer/CommandLineTools/usr/bin/python3"
npm ERR! gyp info spawn /Library/Developer/CommandLineTools/usr/bin/python3
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   '/usr/local/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'make',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas/build/config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/usr/local/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/Users/arn/Library/Caches/node-gyp/16.13.0/include/node/common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=/Users/arn/Library/Caches/node-gyp/16.13.0',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node_modules/npm/node_modules/node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=/Users/arn/Library/Caches/node-gyp/16.13.0/<(target_arch)/node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! /bin/sh: pkg-config: command not found
npm ERR! gyp: Call to 'pkg-config pixman-1 --libs' returned exit status 127 while in binding.gyp. while trying to load binding.gyp
npm ERR! gyp ERR! configure error 
npm ERR! gyp ERR! stack Error: `gyp` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:353:16)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
npm ERR! gyp ERR! System Darwin 20.6.0
npm ERR! gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--module=/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas/build/Release/canvas.node" "--module_name=canvas" "--module_path=/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas/build/Release" "--napi_version=8" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=node-v93"
npm ERR! gyp ERR! cwd /Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas
npm ERR! gyp ERR! node -v v16.13.0
npm ERR! gyp ERR! node-gyp -v v8.2.0
npm ERR! gyp ERR! not ok 
npm ERR! node-pre-gyp ERR! build error 
npm ERR! node-pre-gyp ERR! stack Error: Failed to execute '/usr/local/bin/node /usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --module=/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas/build/Release/canvas.node --module_name=canvas --module_path=/Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas/build/Release --napi_version=8 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v93' (1)
npm ERR! node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/Users/arn/Downloads/hashlips_art_engine-main/node_modules/@mapbox/node-pre-gyp/lib/util/compile.js:89:23)
npm ERR! node-pre-gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
npm ERR! node-pre-gyp ERR! stack     at maybeClose (node:internal/child_process:1064:16)
npm ERR! node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
npm ERR! node-pre-gyp ERR! System Darwin 20.6.0
npm ERR! node-pre-gyp ERR! command "/usr/local/bin/node" "/Users/arn/Downloads/hashlips_art_engine-main/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
npm ERR! node-pre-gyp ERR! cwd /Users/arn/Downloads/hashlips_art_engine-main/node_modules/canvas
npm ERR! node-pre-gyp ERR! node -v v16.13.0
npm ERR! node-pre-gyp ERR! node-pre-gyp -v v1.0.6
npm ERR! node-pre-gyp ERR! not ok

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/arn/.npm/_logs/2021-11-14T13_23_30_413Z-debug.log
arn@As-MacBook-Pro hashlips_art_engine-main %
<img width="775" alt="Screenshot 2021-11-14 at 13 25 59" src="https://user-images.githubusercontent.com/94299025/141683246-c0776074-ccca-4b42-8bf6-9fa78e53f921.png">
 

### Operating system

Mac os big sur

### Scope

Installation

### Module and version

npm install

**Comment by VoltrexMaster** [comment]
Seems like the node-canvas installation is failing for you.

Try running the following:
```bash
$ brew install pkg-config cairo pango libpng jpeg giflib librsvg
```
In your terminal and then run `npm i` again.

If that doesn't work for you, refer to the [compiling section of node-canvas](https://github.com/Automattic/node-canvas#compiling).

**Comment by SuperArchN** [comment]
Thanks so much, just seen this so i will give it a go. 👍🏻 Yeah, homebrew seems to be a popular choice to get around this. I'm pretty annoyed and feel it's m1 chip related, stuff but who knows.

**Comment by SuperArchN** [comment]
I tried it but just get this, annoyingly
zsh: command not found: $

**Comment by jbmagination** [comment]
@SuperArchN just run `brew install pkg-config cairo pango libpng jpeg giflib librsvg`

**Comment by SuperArchN** [comment]
I just get this error when i run it 
 zsh: command not found: brew

Thanks for the reply

**Comment by jbmagination** [comment]
@SuperArchN 

install brew:
`/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`

then try it again:
`brew install pkg-config cairo pango libpng jpeg giflib librsvg`

then try installing

**Comment by SuperArchN** [comment]
Still the same outcome. I had previously installed homebrew but did it again anyway. then tried the brew install pkg-config cairo pango libpng jpeg giflib librsvg again but still get the same 
zsh: command not found: brew

**Comment by jbmagination** [comment]
can you try `/usr/local/bin/brew install pkg-config cairo pango libpng jpeg giflib librsvg`?

if that works, then do `export PATH=/usr/local/bin:$PATH`

**Comment by SuperArchN** [comment]
I just get 
zsh: no such file or directory: /usr/local/bin/brew
Then
zsh: no such file or directory: /Users/homefolder/.zprofile

Thanks for yoiur help

**Comment by jbmagination** [comment]
@SuperArchN
> replace `/Users/homefolder/` with `~/`

**Comment by SuperArchN** [comment]
> `brew install pkg-config cairo pango libpng jpeg giflib librsvg`

Ok , this worked:
echo 'eval $(/opt/homebrew/bin/brew shellenv)' >> ~/.zprofile
eval $(/opt/homebrew/bin/brew shellenv)

And now i am installign the packages via:
brew install pkg-config cairo pango libpng jpeg giflib librsvg


ANd i now have npm installed and Viusal Code is working :-) I can't believe it.
Jbmagination - you are an absolute legend!!!!!!!!
VoltrexMaster - thanks so much for your help too.

Jbmagination, if you don't mind, in layman's terms what was the error ?

**Comment by jbmagination** [comment]
When installing `node-canvas`, it couldn't get a few things needed to install. From what I can tell, you're running a Mac using an Apple processor. You can check and confirm this yourself [here](https://support.apple.com/en-us/HT211814).

Architectures are ways of reading an instruction that is given from a computer (massive oversimplification). Apple chips uses the `arm64` architecture. 

To break down what happened, it attempted to get a download for your type of computer after detecting it.
```
npm ERR! node-pre-gyp info using node@16.13.0 | darwin | arm64
npm ERR! node-pre-gyp http GET https://github.com/Automattic/node-canvas/releases/download/v2.8.0/canvas-v2.8.0-node-v93-darwin-unknown-arm64.tar.gz
```

`node-canvas` does not offer a download for that architecture, so when npm told GitHub "hey, can I get this file?" it correctly said "that doesn't exist".
```
npm ERR! node-pre-gyp ERR! install response status 404 Not Found on https://github.com/Automattic/node-canvas/releases/download/v2.8.0/canvas-v2.8.0-node-v93-darwin-unknown-arm64.tar.gz
```

As such, the source code was downloaded. It attempted to use it to put together the package:
```
npm ERR! Pre-built binaries not installable for canvas@2.8.0 and node@16.13.0 (node-v93 ABI, unknown) (falling back to source compile with node-gyp)
```

However, you didn't have what was needed to install `node-canvas` from source: you needed to install a few things with `brew`.

However, your `brew` installation wasn't done correctly, as the "Next steps" listed weren't done. That meant you couldn't run `brew` to begin with to install the packages.

After the packages were installed, putting it together from the source code worked successfully 🎉

**Comment by SuperArchN** [comment]
I can't credit enough the support you have given me. Thank you so much.
Yeah, the whole issue was to do with the m1. I have node and npm
successfully installed on my desktop (old mac) and that worked first time.

This is brill. Thank you again.

**Comment by batayola1992** [comment]
> `/usr/local/bin/brew install pkg-config cairo pango libpng jpeg giflib librsvg`
you do this in the terminal? Im still getting error on mine .

**Comment by lamriabdellah0** [comment]
This npm install error on M1 Macs, especially with native modules, is such a common hurdle! It really highlights how complex environment setup can be. I've found ServBay dramatically simplifies my entire local dev environment, taking away so many of these dependency and architectural headaches. Great to see the solution here
